Add warp support

smallglitch opened the pull request:
Adds support for the warp framework

Having to add the `.recover()` function is unintuitive but I'm not aware of a better solution  

However, the recover function is not *strictly* necessary. It turns the `Rejection`, which would result in a `500 Internal Server Error`, into a `400 Bad Request`

samscott89 commented:
Hey @smallglitch. Thanks for opening this PR. This seems like a reasonable thing to add, given that we have similar support for actix.

I have a couple of suggestions, let me know what you think: 

- I don't expect many people to need to specify a configuration. If we made the filter a struct wrapping the configuration and implemented `Filter` for it, then people could use that as the default implementation instead?
- Comparing the implementation to warp's equivalent: https://docs.rs/warp/0.3.0/src/warp/filters/query.rs.html#13-27. It would be good for these to be as close as possible. To that end, I think it would be reasonable to have the error default to invalid query (avoiding the need for `recover`. That would be the default). And also, in the case of a missing query string, treating it as the empty string (for entirely optional configuration), which is what `query` does.
- Small nit: could add `tracing` as an optional dep and add a similar debugging line that `warp` does.

How do those sound to you?

samscott89 commented:
Oh! And we should add a simple test. Similar to what we have for actix: https://github.com/samscott89/serde_qs/blob/main/tests/test_actix.rs

smallglitch commented:
Hi, I've added `tracing` debug messages, tests that are applicable and the fallback to an empty string with the last commit

> If we made the filter a struct wrapping the configuration and implemented Filter for it, then people could use that as the default implementation instead?  

I think that's not really possible unfortunately. The `Filter` trait is sealed by requiring the type to implement the private trait `FilterBase`  

> I think it would be reasonable to have the error default to invalid query (avoiding the need for recover. That would be the default)

Warp's `InvalidQuery` struct (`warp::reject::InvalidQuery`) can't be turned into a `Rejection` outside of the `warp` crate itself as far as I know   

(The `warp::query::query()` function uses `warp::reject::invalid_query()` to construct a `Rejection` that resolves to a `400 Bad Request`. Unfortunately this function is private)

The struct seems to only be public to be able to recover from it  
And regular user-defined rejections always result in `500 Internal Server` errors if they're not recovered from  

If we'd try to return a HTTP response ourselves, the filter produce something like a `warp::Either<(T,), (impl warp::Reply,)>` instead of returning immediately. Rejections are the only way to prevent the execution of other `.and()` chained filters

smallglitch commented:
I'm sorry if it's a bad time, I wanted to ping because it's been a while

samscott89 commented:
> 
> 
> I'm sorry if it's a bad time, I wanted to ping because it's been a while

It's all good! Thanks for the reminder, I'll make some time to get to this by the end of the week.

samscott89 commented:
Cool, this looks good. Thanks for adding this!
